Features/mehnaz/resources UI

MEH-NAZZ opened the pull request:
Added a new flow for the resources page

anasbinarif committed:
Merge pull request #35 from anasbinarif/dev

Dev

MEH-NAZZ committed:
new changes in resource page

MEH-NAZZ committed:
added route to dropdown links in navbar

MEH-NAZZ committed:
added new flow for the resources page

MEH-NAZZ committed:
fixed marge conflicts

fizo-neechan commented on src/app/(main)/layout.tsx:
```diff
@@ -19,10 +20,12 @@ export default async function MainLayout({
-        <Navbar />
-        {children}
-        <Footer />
+        <Box sx={{display: 'flex', minHeight: '100vh', flexDirection: 'column'}}>
```
make the variant of this as "main"

fizo-neechan commented on src/components/pagination/Pagination.style.ts:
```diff
@@ -0,0 +1,32 @@
+    color: '#000',
+    fontSize: '14px',
+    fontFamily: 'Lato, sans-serif',
+    [theme.breakpoints.down('sm')]: {
```
For nested styles, please look into the documentation to see if there is a better method to do this

fizo-neechan commented on src/contexts/MultiStepContext.tsx:
```diff
@@ -0,0 +1,123 @@
+  );
+};
+
+export const useMultiStepContext = (): MultiStepContextProps => {
+  const context = useContext(MultiStepContext);
+
+  if (!context) {
+    throw new Error(
+      'useMultiStepContext must be used within a MultiStepProvider'
+    );
+  }
+
+  return context;
+};
```
move this into a separate hooks file

fizo-neechan commented on src/contexts/MultiStepContext.tsx:
```diff
@@ -0,0 +1,123 @@
+  );
+};
+
+export const useMultiStepContext = (): MultiStepContextProps => {
+  const context = useContext(MultiStepContext);
+
+  if (!context) {
+    throw new Error(
+      'useMultiStepContext must be used within a MultiStepProvider'
+    );
+  }
+
+  return context;
+};
```
move this into a separate hooks file

fizo-neechan replied:
```
"use client";
import {useContext} from "react";
import {FormContext} from "../contexts/MultiStepFormContext";

const useMultiStepForm = () => {
  const context = useContext(FormContext);
  if (!context) {
    throw new Error("useForm must be used within a FormProvider");
  }
  return context;
};

export default useMultiStepForm;
```


The Hook should be in a separate file like this

fizo-neechan commented:
@MEH-NAZZ Please follow the PR naming scheme, as well as adding more specific descriptions and screenshots for the PR

fizo-neechan commented on src/features/resources/StepComponent.tsx:
```diff
@@ -0,0 +1,462 @@
+      <Resourceswrapper>
+        <AppContentWrapper>
+          <ResourcesContainer>
+            <Box>
+              <SectionHeading
+                text="Start Practicing"
+                align="center"
+                showLeftLine={true}
+                color="#DA9694"
+                textWidth="150px"
+                gradientType="second"
+              />
+              <ResourcesHeading variant="h2">
+                {currentStep?.stepName}
+              </ResourcesHeading>
+              <Box
+                sx={{
+                  display: 'flex',
+                  alignItems: 'center',
+                  justifyContent: 'center',
+                }}
+              ></Box>
+              <Box
+                sx={{
+                  mt: '30px',
+                  display: { xs: 'none', sm: 'flex' },
+                  alignItems: 'center',
+                }}
+              >
+                <Breadcrumbs
+                  aria-label="breadcrumb"
+                  separator={
+                    <Image
+                      src={'/icons/right-arrow.svg'}
+                      alt={'icon'}
+                      width={9}
+                      height={9}
+                      style={{ transform: 'translateY(20%)' }}
```
\These should be in a layout component and have the specific pages rendered as separate components

fizo-neechan replied:
```

const Index = () => {
    const { theme } = useTheme();
  return (
    <ValidationProvider>
      <FormProvider>
        <Box
          sx={{
            maxWidth: "1170px",
            margin: "0 auto",
            boxShadow: "0px 4px 12px rgba(0, 0, 0, 0.17)",
            borderRadius: "23px",
            padding: "4rem 2rem",
            border: "1px solid #CECECE",
              width: "95%",
              "@media (max-width: 600px)": {
                  boxShadow: "none",
                  border: "none",
                  backgroundColor: "transparent",
                  padding: "2rem 1rem",
                  marginTop: "-5rem",
              },
          }}
        >
            <ServiceToggleContainer>
            <ServiceToggle />
          </ServiceToggleContainer>
          <StepBar />

          <BookingForm />

          <BookingFormFooter />
        </Box>
      </FormProvider>
    </ValidationProvider>
  );
};

export default Index;
```

Here is a sample of the Layout, as you can see the code is calling the `<BookingForm />` component which has the logic for rendering the specific steps according to their step number

fizo-neechan commented on src/features/resources/StepComponent.tsx:
```diff
@@ -0,0 +1,462 @@
+                </ResourcesPara>
+              </Box>
+              <Box>
+                {currentStep?.stepNumber === 5 ? (
+                  <ResourcesTableContainer>
+                    <StyledTable sx={{ minWidth: 768 }}>
+                      <TableHead>
+                        <TableRow>
+                          {Array.isArray(options) &&
+                            options.map((option, index) =>
+                              option?.headings?.map((heading, headingIndex) => (
+                                <TableCell
+                                  key={`${index}-${headingIndex}`}
+                                  sx={{
+                                    fontWeight: 'bold',
+                                    '&:nth-last-child(1)': {
+                                      textAlign: 'right',
+                                    },
+                                  }}
+                                >
+                                  {heading}
+                                </TableCell>
+                              ))
```
As stated above, this should be a separate component. I have shared an example on slack from the fast-clean project, please refer to that and/or reach out if you need more clarity on this implementation

fizo-neechan replied:
```

const BookingForm = () => {
  const { currentStep, formData, nextStep } = useMultiStepForm();

  console.log("BookingForm - Current Step:", currentStep);

  if (currentStep === 1) return <LocationSelection />;
  else if (currentStep === 2) return <LiscencePlateStep />;
  else if (currentStep === 3) return <CarTypeStep />;
  else if (currentStep === 4) return <PackageSelectionStep />;
  else if (currentStep === 5) return <SubscriptionPackagesStep />;
  else if (currentStep === 6 && formData.selectedPackageType === "Subscription Plans") {
    // nextStep();
    return null;
  } else if (currentStep === 6 && formData.selectedPackageType === "Anywhere Autocare") return <AutocarePackagesStep />;
  else if (currentStep === 7) return <AdditionalOptionsStep />;
  else if (currentStep === 8) return <DetailingStep />;
  else if (currentStep === 9) return <SelectCityStep />;
  else if (currentStep === 10) return <ScheduleAppointmentStep />;
  else if (currentStep === 11) return <SummaryStep />;
  else if (currentStep === 12) return <PersonParticularsStep />;

  return <Box>Undefined</Box>;
};

export default BookingForm;
```

Here is some sample code from the Fast Clean project, I'd like you to focus on how all the different components are rendered according to their step number

fizo-neechan commented on src/features/resources/StepComponent.tsx:
```diff
@@ -0,0 +1,462 @@
+                          options.map((option) => (
+                            <>
+                              {option?.data?.map((data, dataIndex) =>
+                                option?.name === 'Revision Notes' ? (
+                                  <TableRow key={dataIndex}>
```
Instead of this, the two types of tables should be different components, and then have them rendered as per the use-case. Though this isn't a big problem but it makes the code-base cleaner and easier to understand down the line

fizo-neechan replied:
This is general best practice, that the code for a specific type of component should be in it's specific component. Since the two types of tables are fundamentally different, so it'd be better if you break them down into their own components

MEH-NAZZ committed:
new implementation for resources pages, created saperate components

MEH-NAZZ commented:
![image](https://github.com/user-attachments/assets/429dfc2a-ea17-42c5-acc3-f07f9167f848)

I have restructured and updated the entire code for the resources page:

- Created separate files for each step and table to keep the code organized.
- Added a MultiStepFormContext to manage the current step, selected options, and functions for navigating steps and setting options.
- Introduced a layout component specifically for the resources page.
- Added a ResourcesStepper.tsx component to control step rendering, displaying each step based on currentStep.
- Created separate components for Breadcrumbs and step navigation to enhance code clarity.
Updated the navbar logic to ensure users are directed to the appropriate resources.
